Userland/Applications/PixelPaint/Tools/GradientTool.cpp

```diff
@@ -153,16 +153,14 @@ void GradientTool::on_keyup(GUI::KeyEvent& event)
     }
 }
 
-void GradientTool::on_second_paint(Layer const* layer, GUI::PaintEvent&)
+void GradientTool::on_second_paint(Layer const* layer, GUI::PaintEvent& event)
 {
     if (!layer || !has_gradient_start_end())
         return;
 
-    // FIXME: Clipping still does overwrite the ruler if we are zoomed in.
-    auto clipping_rect = m_editor->rect().contains(m_editor->content_rect()) ? m_editor->content_rect() : m_editor->rect();
     GUI::Painter painter(*m_editor);
-    painter.add_clip_rect(clipping_rect);
-    draw_gradient(painter, true, m_editor->content_to_frame_position(Gfx::IntPoint(0, 0)), m_editor->scale());
+    painter.add_clip_rect(event.rect());
+    draw_gradient(painter, true, m_editor->content_to_frame_position(Gfx::IntPoint(0, 0)), m_editor->scale(), m_editor->content_rect());
```

Comment:
It seems like `::on_second_paint` knows which clipping rect to set; why not add an `.add_clip_rect` here instead of passing it to `draw_gradient`?

Reply:
Because `draw_gradient()` is drawing both the gradient and the controls, but we only want to clip the gradient and not the controls. So I pass the clip rect so only the gradient is clipped. I think ideally the controls would be drawn separately, but I didn't really want to start refactoring things :sweat_smile: